{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c8f2764b",
   "metadata": {},
   "source": [
    "# Instruction for Prott5 embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6ce4be6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-04-17 17:08:29.405022: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n"
     ]
    }
   ],
   "source": [
    "from embedding import Embedding\n",
    "from Bio import SeqIO\n",
    "import random\n",
    "import copy\n",
    "import joblib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "70bed6f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./data/iter3_blind_protein_prott5.csv'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NOTE WARNING: Do not run this cell if you want to run a Genetic Algorithm. Only run if you want to make embeddings of a specific file\n",
    "# NOTE: it is better to convert this into python file if you don't want to have an interactive section\n",
    "# in embedding.py there is a line warnings.filterwarnings(\"ignore\")\n",
    "# which ignore all the warning when running, if you want to turn that off, comment on that line\n",
    "\n",
    "EMBEDDING_INPUT_FILE  = './data/mutated_iter2.fasta' # input fasta file + its directory\n",
    "EMBEDDING_OUTPUT_NAME = 'iter3_blind' # output file name\n",
    "EMBEDDING_OUTPUT_DIR = \"data\" # output file directory, remember don't put / at the end\n",
    "Embedder = Embedding(in_file=EMBEDDING_INPUT_FILE,\n",
    "                     out_name = EMBEDDING_OUTPUT_NAME, \n",
    "                     out_dir = EMBEDDING_OUTPUT_DIR,\n",
    "                     level='protein', # change to \"protein\" if want to embed in protein level\n",
    "                     embed='prott5')\n",
    "\n",
    "Embedder.embedding()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7244c43b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This code defines a list of amino acids, an exclusion list, and a binding dictionary.\n",
    "It also initializes an empty list for mutated sequences.\n",
    "\n",
    "- amino_acids: A list of amino acids.\n",
    "- exclusion_list: A list of amino acids to be excluded.\n",
    "- binding_dict: A dictionary where the keys are protein identifiers and the values are lists of binding cavity positions.\n",
    "- mutated_sequences: An empty list to store mutated sequences.\n",
    "\"\"\"\n",
    "\n",
    "amino_acids = ['A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V'] # XZBJ added \n",
    "exclusion_list = ['B', 'Z', 'J', 'O', 'U', 'X'] # XZBJ added\n",
    "binding_dict = {\n",
    "    'gi|932248239|gb|ALG62950.1|': [116, 290, 291, 292, 308, 309, 310, 311, 312, 313, 315, 316, 317, 318, 319, 320, 321],\n",
    "    'gi|932248233|gb|ALG62947.1|': [116, 290, 291, 292, 308, 309, 310, 311, 312, 313, 315, 316, 317, 318, 319, 320, 321],\n",
    "    'gi|932248235|gb|ALG62948.1|': [116, 290, 291, 292, 308, 309, 310, 311, 312, 313, 315, 316, 317, 318, 319, 320, 321],\n",
    "    'gi|932247944|gb|ALG62808.1|': [116, 290, 291, 292, 308, 309, 310, 311, 312, 313, 315, 316, 317, 318, 319, 320, 321],\n",
    "    'gi|932247942|gb|ALG62807.1|': [116, 290, 291, 292, 308, 309, 310, 311, 312, 313, 315, 316, 317, 318, 319, 320, 321],\n",
    "    'gi|932247938|gb|ALG62805.1|': [116, 290, 291, 292, 308, 309, 310, 311, 312, 313, 315, 316, 317, 318, 319, 320, 321],\n",
    "    'gi|932247946|gb|ALG62809.1|': [167, 168, 169, 185, 186, 187, 188, 189, 190, 192, 193, 194, 195, 196, 197, 198],\n",
    "    'gi|932248269|gb|ALG62965.1|': [116, 290, 291, 292, 308, 309, 310, 311, 312, 313, 315, 316, 317, 318, 319, 320, 321],\n",
    "    'gi|313473685|dbj|BAJ40208.1|': [111, 285, 286, 287, 303, 304, 305, 306, 307, 308, 310, 311, 312, 313, 314, 315, 316],\n",
    "    'gi|932247975|gb|ALG62823.1|': [116, 290, 291, 292, 308, 309, 310, 311, 312, 313, 315, 316, 317, 318, 319, 320, 321],\n",
    "    'AFB70630.1': [135, 309, 310, 311, 327, 328, 329, 330, 331, 332, 334, 335, 336, 337, 338, 339, 340],\n",
    "    'tr|A0A1C3HPM4|A0A1C3HPM4_9POAL': [135, 309, 310, 311, 327, 328, 329, 330, 331, 332, 334, 335, 336, 337, 338, 339, 340],\n",
    "    'tr|A0A1C3HPS9|A0A1C3HPS9_PUCDI': [135, 309, 310, 311, 327, 328, 329, 330, 331, 332, 334, 335, 336, 337, 338, 339, 340],\n",
    "    'tr|A0A1C3HPT0|A0A1C3HPT0_9POAL': [135, 309, 310, 311, 327, 328, 329, 330, 331, 332, 334, 335, 336, 337, 338, 339, 340],\n",
    "    'tr|A0A6C0SV93|A0A6C0SV93_ERATE': [142, 316, 317, 318, 334, 335, 336, 337, 338, 339, 341, 342, 343, 344, 345, 346, 347],\n",
    "    'YP_899415.1': [135, 309, 310, 311, 327, 328, 329, 330, 331, 332, 334, 335, 336, 337, 338, 339, 340],\n",
    "    'SCM15160.1': [135, 309, 310, 311, 327, 328, 329, 330, 331, 332, 334, 335, 336, 337, 338, 339, 340],\n",
    "    'SCM15158.1': [135, 309, 310, 311, 327, 328, 329, 330, 331, 332, 334, 335, 336, 337, 338, 339, 340],\n",
    "    'YP_009573569.1': [142, 316, 317, 318, 334, 335, 336, 337, 338, 339, 341, 342, 343, 344, 345, 346, 347],\n",
    "    'AGT56139.1': [121, 295, 296, 297, 313, 314, 315, 316, 317, 318, 320, 321, 322, 323, 324, 325, 326]\n",
    "}\n",
    "mutated_sequences = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f75410f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def start_embedding(iter, inputFile=None, j=-1, temp=False):\n",
    "    \"\"\"\n",
    "    Starts the embedding process.\n",
    "\n",
    "    Parameters:\n",
    "    - iter (int): The iteration number.\n",
    "    - inputFile (str, optional): The input file path. If not provided, a default file path will be used.\n",
    "    - j (int, optional): The value of j. If -1, a default value will be used.\n",
    "    - temp (bool, optional): Whether to use a temporary output name. Default is False.\n",
    "\n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "\n",
    "    if j != -1:\n",
    "        EMBEDDING_INPUT_FILE = f'./data/mutated_iter{iter}-{j}.fasta'  # input fasta file + its directory\n",
    "    else:\n",
    "        EMBEDDING_INPUT_FILE = f'./data/mutated_iter{iter}.fasta'  # input fasta file + its directory\n",
    "\n",
    "    if inputFile is not None:\n",
    "        EMBEDDING_INPUT_FILE = inputFile\n",
    "\n",
    "    if temp:\n",
    "        EMBEDDING_OUTPUT_NAME = f'temp_blind'\n",
    "    elif j != -1:\n",
    "        EMBEDDING_OUTPUT_NAME = f'iter{iter+1}-{j}_blind'\n",
    "    else:\n",
    "        EMBEDDING_OUTPUT_NAME = f'iter{iter+1}_blind'  # output file name\n",
    "    EMBEDDING_OUTPUT_DIR = \"data\"  # output file directory, remember don't put / at the end\n",
    "    Embedder = Embedding(\n",
    "        in_file=EMBEDDING_INPUT_FILE,\n",
    "        out_name=EMBEDDING_OUTPUT_NAME,\n",
    "        out_dir=EMBEDDING_OUTPUT_DIR,\n",
    "        level='protein',  \n",
    "        embed='prott5'\n",
    "    )\n",
    "\n",
    "    Embedder.embedding()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d9e7882b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def single_replacement(sequence, iterations):\n",
    "    \"\"\"\n",
    "    Perform single base replacement mutation on a given sequence. Here we randomly select a position in the sequence and replace the base at that position with a random base.\n",
    "\n",
    "    Args:\n",
    "        sequence (Bio.SeqRecord.SeqRecord): The input sequence to mutate.\n",
    "        iterations (int): The number of iterations to perform the mutation.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of mutated sequences.\n",
    "\n",
    "    \"\"\"\n",
    "    result = []\n",
    "    data = sequence.seq\n",
    "    desc = sequence.description\n",
    "\n",
    "    seq_id = sequence.id\n",
    "    binding_indices = binding_dict[seq_id]\n",
    "\n",
    "    \n",
    "    new_seq = sequence\n",
    "    for i in range(iterations):\n",
    "        new_seq = copy.deepcopy(sequence)\n",
    "        \n",
    "        # pick a random position (element) from the binding indices\n",
    "        pos = random.choice(binding_indices)\n",
    "        # amino acid list without the current amino acid\n",
    "\n",
    "        removed_aa = amino_acids.copy()\n",
    "        if data[pos] not in ['X' ,'Z', 'B','J']:\n",
    "            removed_aa.remove(data[pos])\n",
    "        # pick a random base\n",
    "        new_base = random.choice(removed_aa)\n",
    "        # replace the base at that position\n",
    "        # print(\"Replacing\", data[pos], \"with\", new_base, \"at index\", pos)\n",
    "        mutated_data = data[:pos] + new_base + data[pos+1:]\n",
    "\n",
    "        new_desc = desc + f\"|_{i+1}-SR-{data[pos]}{pos}{new_base}\"\n",
    "        new_seq.seq = mutated_data\n",
    "        new_seq.description = new_desc\n",
    "        result.append(new_seq)\n",
    "    \n",
    "    return result\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1db8d36e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def double_replacement(sequence, iterations):\n",
    "    \"\"\"\n",
    "    Perform double replacement mutation on a given sequence. Here we randomly select two positions in the sequence and replace the bases at those positions with random bases.\n",
    "\n",
    "    Args:\n",
    "        sequence (Bio.SeqRecord.SeqRecord): The input sequence to mutate.\n",
    "        iterations (int): The number of iterations to perform the mutation.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of mutated sequences.\n",
    "\n",
    "    \"\"\"\n",
    "    result = []\n",
    "    data = sequence.seq\n",
    "    desc = sequence.description\n",
    "    seq_id = sequence.id\n",
    "    binding_indices = binding_dict[seq_id]\n",
    "\n",
    "    new_seq = sequence\n",
    "    for i in range(iterations):\n",
    "        new_seq = copy.deepcopy(sequence)\n",
    "        new_desc = desc + f\"|_{i+1}-DR\"\n",
    "        # pick a random position (element) from the binding indices\n",
    "        pos1 = random.choice(binding_indices)\n",
    "        pos2 = random.choice(binding_indices)\n",
    "        while pos1 == pos2:\n",
    "            pos2 = random.choice(binding_indices)\n",
    "\n",
    "        for pos in [pos1, pos2]:\n",
    "            # amino acid list without the current amino acid\n",
    "            removed_aa = amino_acids.copy()\n",
    "            if data[pos] not in ['X' ,'Z', 'B','J']:\n",
    "                removed_aa.remove(data[pos])\n",
    "            # pick a random base\n",
    "            new_base = random.choice(removed_aa)\n",
    "            # replace the base at that position\n",
    "    \n",
    "            #print(\"Replacing\", data[pos], \"with\", new_base, \"at index\", pos)\n",
    "            new_desc += f\"-{data[pos]}{pos}{new_base}\"\n",
    "            mutated_data = data[:pos] + new_base + data[pos+1:]\n",
    "\n",
    "        new_seq.seq = mutated_data\n",
    "        new_seq.description = new_desc\n",
    "        result.append(new_seq)\n",
    "\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9a89fa8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def swap(sequence, iterations):\n",
    "    \"\"\"\n",
    "    Swaps elements in a sequence based on the given iterations. Here we randomly select two positions in the sequence and swap the elements at those positions.\n",
    "\n",
    "    Args:\n",
    "        sequence (Sequence): The input sequence to swap elements in.\n",
    "        iterations (int): The number of iterations to perform the swapping.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of sequences with swapped elements.\n",
    "    \"\"\"\n",
    "    result = []\n",
    "    data = sequence.seq\n",
    "    desc = sequence.description\n",
    "    seq_id = sequence.id\n",
    "    binding_indices = binding_dict[seq_id]\n",
    "    new_seq = sequence\n",
    "    for i in range(iterations):\n",
    "        new_seq = copy.deepcopy(sequence)\n",
    "        new_data = data\n",
    "        # pick a random position (element) from the binding indices\n",
    "        pos1 = random.choice(binding_indices)\n",
    "        pos2 = random.choice(binding_indices)\n",
    "        while pos1 == pos2:\n",
    "            pos2 = random.choice(binding_indices)\n",
    "        #print(\"Swapping\", new_data[pos1], \"and\", new_data[pos2], \"at indices\", pos1, \"and\", pos2)\n",
    "        new_desc = desc + f\"|_{i+1}-S-{new_data[pos1]}{pos1}{new_data[pos2]}{pos2}\"\n",
    "        new_data = new_data[:pos1] + new_data[pos2] + new_data[pos1+1:]\n",
    "        new_data = new_data[:pos2] + new_data[pos1] + new_data[pos2+1:]\n",
    "\n",
    "        new_seq.seq = new_data\n",
    "        new_seq.description = new_desc\n",
    "        result.append(new_seq)\n",
    "\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ffe8baa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_sequences(iterations, file, j=-1):\n",
    "    \"\"\"\n",
    "    Mutates the sequences in a FASTA file by performing single replacement, double replacement, and swapping operations. The mutated sequences are written to a new FASTA file.\n",
    "\n",
    "    Parameters:\n",
    "    iterations (int): The number of iterations for mutation.\n",
    "    file (str): The path to the input FASTA file.\n",
    "    j (int, optional): The value used for creating the output file name. Defaults to -1.\n",
    "\n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "    mutated_sequences = []  # empty the list\n",
    "    fasta_file = file\n",
    "    # Read the FASTA file\n",
    "    sequences = SeqIO.parse(fasta_file, \"fasta\")\n",
    "\n",
    "    for sequence in sequences:\n",
    "        # Access the sequence ID and sequence data\n",
    "        sequence_desc = sequence.description\n",
    "        sequence_data = sequence.seq\n",
    "\n",
    "        mutated_sequences.extend(single_replacement(sequence, 20))\n",
    "        mutated_sequences.extend(double_replacement(sequence, 20))\n",
    "        mutated_sequences.extend(swap(sequence, 10))\n",
    "\n",
    "    # Specify the path for the new FASTA file\n",
    "    if j == -1:\n",
    "        output_file = f\"./data/mutated_iter{iterations}.fasta\"\n",
    "    else:\n",
    "        output_file = f\"./data/mutated_iter{iterations}-{j}.fasta\"\n",
    "\n",
    "    # Write the sequences to the new FASTA file\n",
    "    SeqIO.write(mutated_sequences, output_file, \"fasta\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "cd60ee9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def grab_top(iterations, number, multi):\n",
    "    \"\"\"\n",
    "    Retrieves the top proteins based on predictions of Kcat values. \n",
    "\n",
    "    Args:\n",
    "        iterations (int): The number of iterations.\n",
    "        number (int): The number of top proteins to retrieve.\n",
    "        multi (str): The multi value.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of the top proteins based on Kcat predictions.\n",
    "    \"\"\"\n",
    "    df = pd.read_csv(f\"./data/iter{iterations+1}-{multi}_blind_protein_prott5.csv\")\n",
    "    proteinId = df['ProteinID']\n",
    "    embeddings = df.drop(columns=['ProteinID'])\n",
    "    kcat = joblib.load(r\"./ridgeModels/kcat mean 0.01.pkl\")\n",
    "    sco = joblib.load(r\"./ridgeModels/Sco mean 0.01.pkl\")\n",
    "    kcat_predictions = kcat.predict(embeddings)\n",
    "    sco_predictions = sco.predict(embeddings)\n",
    "\n",
    "\n",
    "    kcat_predictions = pd.DataFrame(kcat_predictions, columns=['Kcat'])\n",
    "    sco_predictions = pd.DataFrame(sco_predictions, columns=['Sc/o'])\n",
    "    new_df = pd.concat([proteinId, kcat_predictions, sco_predictions], axis=1)\n",
    "    new_df.to_csv(f\"./data/Iter-{iterations+1}-{multi}-predictions.csv\", index=False)\n",
    "    print(new_df.shape)\n",
    "    top_20 = new_df.nlargest(number, 'Kcat')\n",
    "    top_20 = top_20['ProteinID']\n",
    "    top_20 = top_20.to_list()\n",
    "    top_20 = [string.replace(\">\", \"\") for string in top_20]\n",
    "    \n",
    "    return top_20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d4bc0cdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def grab_OG_sep_top(iterations, number, multi=-1):\n",
    "    \"\"\"\n",
    "    Retrieves the top protein sequences based on predictions of Kcat and Sc/o values. but this function is used to force the original parent sequences to be included in the top proteins.\n",
    "\n",
    "    Parameters:\n",
    "    iterations (int): The number of iterations.\n",
    "    number (int): The number of top protein sequences to retrieve.\n",
    "    multi (int, optional): The multi value. Defaults to -1.\n",
    "\n",
    "    Returns:\n",
    "    list: A list of top protein sequences.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    if multi == -1:\n",
    "        df = pd.read_csv(f\"./data/iter{iterations+1}_blind_protein_prott5.csv\")\n",
    "    else:\n",
    "        df = pd.read_csv(f\"./data/iter{iterations+1}-{multi}_blind_protein_prott5.csv\")\n",
    "    proteinId = df['ProteinID']\n",
    "    print(proteinId.shape)\n",
    "\n",
    "    top_ids = ['gi|313473685|dbj|BAJ40208.1|', 'gi|932247975|gb|ALG62823.1|', 'gi|932248269|gb|ALG62965.1|', 'gi|932248239|gb|ALG62950.1|', 'gi|932247944|gb|ALG62808.1|', 'gi|932247942|gb|ALG62807.1|', 'gi|932247938|gb|ALG62805.1|', 'gi|932248235|gb|ALG62948.1|', 'gi|932248233|gb|ALG62947.1|', 'gi|932247946|gb|ALG62809.1|', 'tr|A0A1C3HPS9|A0A1C3HPS9_PUCDI', 'tr|A0A1C3HPT0|A0A1C3HPT0_9POAL', 'YP_009573569.1', 'AFB70630.1', 'AGT56139.1', 'YP_899415.1', 'SCM15160.1', 'SCM15158.1', 'tr|A0A1C3HPM4|A0A1C3HPM4_9POAL', 'tr|A0A6C0SV93|A0A6C0SV93_ERATE']\n",
    "    kcat = joblib.load(r\"./ridgeModels/kcat mean 0.01.pkl\")\n",
    "    sco = joblib.load(r\"./ridgeModels/Sco mean 0.01.pkl\")\n",
    "\n",
    "    top_sequences = []\n",
    "    print('running Og grab')\n",
    "    for i in top_ids:\n",
    "        new_df = pd.DataFrame()\n",
    "        for j in proteinId: \n",
    "            if i in j:\n",
    "                data = df.loc[df['ProteinID'] == j]\n",
    "                new_df = pd.concat([new_df, data], ignore_index=True)\n",
    "\n",
    "\n",
    "        embeddings = new_df.drop(['ProteinID'], axis=1)\n",
    "        \n",
    "        kcat_predictions = kcat.predict(embeddings)\n",
    "        sco_predictions = sco.predict(embeddings)\n",
    "\n",
    "        kcat_predictions = pd.DataFrame(kcat_predictions, columns=['Kcat'])\n",
    "        sco_predictions = pd.DataFrame(sco_predictions, columns=['Sc/o'])\n",
    "        new_pred_df = pd.concat([new_df['ProteinID'], kcat_predictions, sco_predictions], axis=1)\n",
    "        \n",
    "        print(new_pred_df.shape)\n",
    "        top_20 = new_pred_df.nlargest(number, 'Kcat') # top x number proteins with highest Kcat\n",
    "        top_20 = top_20['ProteinID']\n",
    "        top_20 = top_20.to_list()\n",
    "        top_20 = [string.replace(\">\", \"\") for string in top_20]\n",
    "        \n",
    "        # add to top_sequences\n",
    "        top_sequences.extend(top_20)\n",
    "\n",
    "\n",
    "    embeddings = df.drop(columns=['ProteinID'])\n",
    "    kcat_predictions = kcat.predict(embeddings)\n",
    "    sco_predictions = sco.predict(embeddings)\n",
    "\n",
    "\n",
    "    kcat_predictions = pd.DataFrame(kcat_predictions, columns=['Kcat'])\n",
    "    sco_predictions = pd.DataFrame(sco_predictions, columns=['Sc/o'])\n",
    "    new_df =pd. DataFrame()\n",
    "    new_df = pd.concat([proteinId, kcat_predictions, sco_predictions], axis=1)\n",
    "\n",
    "    # save the predictions of every protein\n",
    "    if multi >= 0:\n",
    "        new_df.to_csv(f\"./data/Iter-{iterations+1}-predictions-{multi}.csv\", index=False)\n",
    "    else:\n",
    "        new_df.to_csv(f\"./data/Iter-{iterations+1}-predictions.csv\", index=False)\n",
    "\n",
    "    return top_sequences\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a4ed4124",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_top_seq_fasta(iterations, top_20, multi=-1):\n",
    "    \"\"\"\n",
    "    Save the top sequences from a FASTA file to a new FASTA file.\n",
    "\n",
    "    Parameters:\n",
    "    - iterations (int): The iteration number.\n",
    "    - top_20 (list): A list of sequence IDs representing the top sequences.\n",
    "    - multi (int, optional): The multi number. Defaults to -1.\n",
    "\n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "    # Specify the path to the FASTA file\n",
    "    if multi == -1:\n",
    "        fasta_file = f\"./data/mutated_iter{iterations}.fasta\"\n",
    "    else:\n",
    "        fasta_file = f\"./data/mutated_iter{iterations}-{multi}.fasta\"\n",
    "\n",
    "    # Read the FASTA file\n",
    "    sequences = SeqIO.parse(fasta_file, \"fasta\")\n",
    "\n",
    "    top_sequences = []\n",
    "    # Iterate over the sequences\n",
    "    for sequence in sequences:\n",
    "        # Access the sequence ID and sequence data\n",
    "        sequence_desc = sequence.description\n",
    "        sequence_data = sequence.seq\n",
    "        if sequence_desc not in top_20:\n",
    "            continue\n",
    "        # Do something with the sequence ID and sequence data\n",
    "        #print(f\"Sequence ID: {sequence_desc}\")\n",
    "        #print(f\"Sequence Data: {sequence_data}\")\n",
    "        top_sequences.append(sequence)\n",
    "\n",
    "    # Specify the path for the new FASTA file\n",
    "    if multi>=0:\n",
    "        output_file = f\"./data/top_sequences_iter{iterations+1}-{multi}.fasta\"\n",
    "    else:\n",
    "        output_file = f\"./data/top_sequences_iter{iterations+1}.fasta\"\n",
    "\n",
    "    # Write the sequences to the new FASTA file\n",
    "    SeqIO.write(top_sequences, output_file, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8c77b4e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "(100, 3)\n",
      "1\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "2\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "3\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "4\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "5\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "6\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "7\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "8\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "9\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "2\n",
      "(100, 3)\n",
      "1\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "2\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "3\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "4\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "5\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "6\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "7\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "8\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "9\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "3\n",
      "(100, 3)\n",
      "1\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "2\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "3\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "4\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "5\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "6\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "7\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "8\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "9\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "4\n",
      "(100, 3)\n",
      "1\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "2\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "3\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "4\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "5\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "6\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "7\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "8\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "9\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "5\n",
      "(100, 3)\n",
      "1\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "2\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "3\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "4\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "5\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "6\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "7\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "8\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "9\n",
      "(5000,)\n",
      "running Og grab\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n",
      "(250, 3)\n"
     ]
    }
   ],
   "source": [
    "# NOTE TLDR: Runs Genetic Algorithm to find the best sequences (Only mutates in the cavity regions)\n",
    "\"\"\"\n",
    "This code performs a series of operations on protein sequences to generate mutated sequences and make predictions using pre-trained models.\n",
    "The steps involved are as follows:\n",
    "1. Load pre-trained models for kcat and sco using joblib.\n",
    "2. Iterate over a range of values from 1 to 5.\n",
    "3. Read protein sequences from a FASTA file.\n",
    "4. Generate mutated sequences by performing single replacement, double replacement, and swap operations on the original sequences.\n",
    "5. Write the mutated sequences to a temporary FASTA file.\n",
    "6. Perform embedding on the temporary FASTA file using a start_embedding function.\n",
    "7. Read the resulting embeddings from a CSV file.\n",
    "8. Make predictions for kcat and sco using the pre-trained models.\n",
    "9. Combine the predictions with the protein IDs and save the results to a CSV file.\n",
    "10. Select the top 5 sequences based on the kcat predictions.\n",
    "11. Filter the mutated sequences to include only the top 5 sequences.\n",
    "12. Write the filtered sequences to a new FASTA file.\n",
    "13. Perform embedding on the filtered sequences using a start_embedding function.\n",
    "14. Select the top 100 sequences based on the embeddings.\n",
    "15. Save the top 100 sequences to a new FASTA file.\n",
    "16. Iterate over a range of values from 1 to 10.\n",
    "17. Mutate the top sequences from the previous step.\n",
    "18. Perform embedding on the mutated sequences.\n",
    "19. Select the top 5 sequences based on the embeddings.\n",
    "20. Save the top 5 sequences to a new FASTA file.\n",
    "\"\"\"\n",
    "\n",
    "runs = 5  # number of runs of GA (Change this if you want to run more or less GA's)\n",
    "iters = 10 # number of iterations for each GA (Change this if you want to run more or less iterations)\n",
    "\n",
    "\n",
    "kcat = joblib.load(r\"./ridgeModels/kcat mean 0.01.pkl\")\n",
    "sco = joblib.load(r\"./ridgeModels/Sco mean 0.01.pkl\")\n",
    "\n",
    "for j in range(1,runs+1):\n",
    "    print(j)\n",
    "    # Load models using pickle instead of joblib\n",
    "    iterations = 0\n",
    "\n",
    "    mutated_sequences = [] #empty the list\n",
    "    temp_sequences = [] #empty the list\n",
    "\n",
    "    fasta_file = f\"./data/top_sequences.fasta\"\n",
    "    # Read the FASTA file\n",
    "    sequences = SeqIO.parse(fasta_file, \"fasta\")\n",
    "\n",
    "    for sequence in sequences:\n",
    "        temp_sequences = []\n",
    "        # Access the sequence ID and sequence data\n",
    "        sequence_desc = sequence.description\n",
    "        sequence_data = sequence.seq\n",
    "\n",
    "        temp_sequences.append(sequence)\n",
    "\n",
    "        temp_sequences.extend(single_replacement(sequence,20))\n",
    "\n",
    "        temp_sequences.extend(double_replacement(sequence,20))\n",
    "\n",
    "        temp_sequences.extend(swap(sequence,10))\n",
    "\n",
    "        # Write the sequences to the new FASTA file\n",
    "        SeqIO.write(temp_sequences, './data/tempfile.fasta', \"fasta\")\n",
    "\n",
    "        start_embedding(iterations, './data/tempfile.fasta', -1, True)\n",
    "\n",
    "        df = pd.read_csv(f\"./data/temp_blind_protein_prott5.csv\")\n",
    "        proteinId = df['ProteinID']\n",
    "        embeddings = df.drop(columns=['ProteinID'])\n",
    "        kcat_predictions = kcat.predict(embeddings)\n",
    "        sco_predictions = sco.predict(embeddings)\n",
    "\n",
    "        kcat_predictions = pd.DataFrame(kcat_predictions, columns=['Kcat'])\n",
    "        sco_predictions = pd.DataFrame(sco_predictions, columns=['Sc/o'])\n",
    "        new_df = pd.concat([proteinId, kcat_predictions, sco_predictions], axis=1)\n",
    "        new_df.to_csv(f\"./data/Iter-{iterations+1}-{j}-predictions.csv\", index=False)\n",
    "\n",
    "        top_5 = new_df.nlargest(5, 'Kcat')\n",
    "        top_5 = top_5['ProteinID']\n",
    "        top_5 = top_5.to_list()\n",
    "        top_5 = [string.replace(\">\", \"\") for string in top_5]\n",
    "\n",
    "\n",
    "        # Specify the path to the FASTA file\n",
    "        temp_fasta_file = f\"./data/tempfile.fasta\"\n",
    "\n",
    "        # Read the FASTA file\n",
    "        tsequences = SeqIO.parse(temp_fasta_file, \"fasta\")\n",
    "\n",
    "        # Iterate over the sequences\n",
    "        for tsequence in tsequences:\n",
    "            # Access the sequence ID and sequence data\n",
    "            sequence_desc = tsequence.description\n",
    "            sequence_data = tsequence.seq\n",
    "            if sequence_desc not in top_5:\n",
    "                continue\n",
    "            mutated_sequences.append(tsequence)\n",
    "\n",
    "    output_file = f\"./data/mutated_iter{iterations}-{j}.fasta\"\n",
    "\n",
    "    # Write the sequences to the new FASTA file\n",
    "    SeqIO.write(mutated_sequences, output_file, \"fasta\")\n",
    "\n",
    "    start_embedding(iterations, j=j)\n",
    "\n",
    "    top_100 = grab_top(iterations, 100, j)\n",
    "\n",
    "    save_top_seq_fasta(iterations, top_100, j)\n",
    "\n",
    "    for i in range(1,iters+1):\n",
    "        print(i)\n",
    "        mutate_sequences(i, f\"./data/top_sequences_iter{i}-{j}.fasta\", j)\n",
    "\n",
    "        start_embedding(i, f'./data/mutated_iter{i}-{j}.fasta', j = j)\n",
    "\n",
    "        top_seq = grab_OG_sep_top(i, 5, j)\n",
    "\n",
    "        save_top_seq_fasta(i, top_seq, j)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
